pkg/buses/message_dispatcher_test.go

```diff
@@ -0,0 +1,344 @@
+			},
+			expectedDestRequest: &requestValidation{
+				url: "http://test-destination-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"id123"},
+					"knative-1":    {"knative-1-value"},
+					"knative-2":    {"knative-2-value"},
+					"ce-abc":       {"ce-abc-value"},
+				},
+				body: "destination",
+			},
+			fakeResponse: &http.Response{
+				StatusCode: http.StatusAccepted,
+				Header: map[string][]string{
+					"do-not-passthrough": {"no"},
+					"x-request-id":       {"altered-id"},
+					"knative-1":          {"new-knative-1-value"},
+					"ce-abc":             {"new-ce-abc-value"},
+				},
+				Body: ioutil.NopCloser(bytes.NewBufferString("destination-response")),
+			},
+			expectedReplyRequest: &requestValidation{
+				url: "http://test-reply-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"altered-id"},
+					"knative-1":    {"new-knative-1-value"},
+					"ce-abc":       {"new-ce-abc-value"},
+				},
+				body: "destination-response",
+			},
+		},
+	}
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			md := NewMessageDispatcher(zap.NewNop().Sugar())
+			fc := &fakeHttpClient{
+				response: tc.fakeResponse,
+			}
+			md.httpClient = fc
+			err := md.DispatchMessage(tc.message, tc.destination, tc.replyTo, DispatchDefaults{})
+			if tc.expectedErr != (err != nil) {
+				t.Errorf("Unexpected error from DispatchRequest. Expected %v. Actual: %v", tc.expectedErr, err)
+			}
+			if tc.expectedDestRequest != nil {
+				rv := fc.popRequest(t)
+				assertEquality(t, *tc.expectedDestRequest, rv)
+			}
+			if tc.expectedReplyRequest != nil {
+				rv := fc.popRequest(t)
+				assertEquality(t, *tc.expectedReplyRequest, rv)
+			}
+			if len(fc.requests) != 0 {
+				t.Errorf("Unexpected requests: %+v", fc.requests)
+			}
+		})
+	}
+}
+
+type requestValidation struct {
+	url     string
+	headers http.Header
+	body    string
+}
+
+type fakeHttpClient struct {
+	t        *testing.T
+	response *http.Response
+	requests []requestValidation
+}
+
+var _ httpDoer = &fakeHttpClient{}
+
+func (f *fakeHttpClient) Do(r *http.Request) (*http.Response, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		f.t.Error("Failed to read the request body")
+	}
+	f.requests = append(f.requests, requestValidation{
+		url:     r.URL.String(),
+		headers: r.Header,
+		body:    string(body),
+	})
+	if f.response != nil {
+		return f.response, nil
+	}
+	return &http.Response{
+		StatusCode: http.StatusAccepted,
+		Body:       ioutil.NopCloser(bytes.NewBufferString("body")),
+	}, nil
+}
+
+func (f *fakeHttpClient) popRequest(t *testing.T) requestValidation {
+	if len(f.requests) == 0 {
+		t.Error("Unable to pop request")
+	}
+	rv := f.requests[0]
+	f.requests = f.requests[1:]
+	return rv
+}
+
+func assertEquality(t *testing.T, expected, actual requestValidation) {
+	if diff := cmp.Diff(expected.url, actual.url); diff != "" {
```

Comment:
i think you can just compare the entire object all in one go, not breaking it apart.